### ISSUE TITLE
Umbrella projects

### ISSUE THREAD

**Issue by eproxus** [bitwalker/exrm#116]
That `exrm` generates one release for each app in an umbrella projekt is very strange. The whole point of an umbrella project is to have a repository of several apps that end up in one system (Erlang node).

The default for generating a release of an umbrella project should be one top level release with all applications in `apps` included.

**Comment by miwee** [comment]
Please refer to https://github.com/bitwalker/exrm-umbrella-test
Having this explicit behaviour is better, as there are dependecies in application order and generating common configuration files, which can be better handled by user directly. Plus as a bonus, if in production system one needs to do some debug, measurements, tuning etc., then it can be tried individually running each app in its own vm.

**Comment by eproxus** [comment]
For a system of say seven apps in an umbrella project, generating seven "pyramid" releases is probably 1% of use cases whereas generating one release based on the "top of the pyramid" app (including all other apps as dependencies) is probably 99% of uses cases for umbrella projects.

Based on my experience, umbrella projects in both Erlang and Elixir is mostly about creating a runtime system that is composed of several apps instead of one. However, the apps are not different or independent enough to warrant separate repositories. That is, they will in 99% of cases all run together, it is just a way of organizing code.

There might be the rare case when you want to run one of them standalone in which it makes sense to pull it out as a dependency instead it being of an umbrella app. That case could still be handled by special flags to exrm, like `mix release thatapp` or something to that effect.

In summary, building one release per app in an umbrella project is untypical and a waste, because very few of those releases will ever be used.

**Comment by liveforeverx** [comment]
@eproxus being little bit late, most `umbrella` projects, which I have seen were meta-projects, which have included all projects as dependency.

**Comment by eproxus** [comment]
@liveforeverx Then there would have been no need to create an umbrella project in the first place? The point of an umbrella project is to bundle several apps in the same repo, not as dependencies.

**Comment by bitwalker** [comment]
@eproxus When I last worked on umbrella support, from what I could tell there wasn't a way to generate such a release. Do you have thoughts on how that should work? Within an umbrella project, each app is really it's own standalone project with it's own `mix.exs`, dependencies, etc. How would you roll those up into a new application at top level? I imagine there is a solution, but I haven't had many (well any other than this thread) people request that feature, so I haven't invested any energy in to it, but I can understand the use case. I'm open to suggestions.

**Comment by eproxus** [comment]
@bitwalker The traditional way to look at Erlang OTP applications, is as libraries (that can include modules, process hierarchies etc.). A release is a combination of OTP applications that work together and are dependent on each other (in most cases). As OTP applications can be more general, they can be included in different releases.

```
+-----release1------+   +-----release2------+
| +------+ +------+ |   | +------+ +------+ |
| | app1 | | app2 | |   | | app3 | | app4 | |
| +------+ +------+ |   | +------+ +------+ |
|                   |   |                   |
|        +---------shared----------+        |
|        |  +------+     +------+  |        |
|        |  | lib1 |     | lib2 |  |        |
|        |  +------+     +------+  |        |
|        +----------+---+----------+        |
|                   |   |                   |
+-------------------+   +-------------------+
```
Typically, in such a situation (no matter if you use Rebar or Mix), you would put `lib1` and `lib2` in their own repos and pull them in as dependencies. `release1` would be an umbrella project containing `app1` and `app2` as apps (and the same for `release2`) which would then depend on the shared apps.

**Comment by bitwalker** [comment]
@eproxus I think you're missing my actual point, which is that with the way things work now, the umbrella project (`release1` in your example) does not actually represent an application in Elixir. I mean, logically/organizationally it does, but compiling it produces no artifact, instead the children apps are built entirely independently. The issue I'm trying to raise is that in order for a release to be produced at the top level (`release1` for example), we have to merge the children together somehow. Creating the .rel for `release1` is the easy part as far as I can tell, the questions around how to handle configuration, upgrades/appups, etc. become much more complicated I fear.

**Comment by eproxus** [comment]
In the Erlang (and Elixir) world, a release is built from a collection of apps. There are no top level artifacts except for configuration. Mix already supports the top level `config` folder for umbrella projects, which is all you would need (`apps`, `deps` and `config`).

This is the way it should be done in the Erlang world, and in the Elixir world as well. Look at [Relx](https://github.com/erlware/relx) for a good example. Rebar3 is also [working on this](http://www.rebar3.org/v3.0/docs/releases) using Relx.

**Comment by bitwalker** [comment]
I make use of relx in exrm, and am quite familiar with it. I know it's possible to do what you are asking, that's not really the issue, in fact I'm not really saying there is an issue, it's simply a fact that this would be rethinking how we do umbrella apps entirely in exrm (and probably Elixir as a community really when it comes to releases). Perhaps that's a good thing, but I'd like to hear from as many people who are using umbrella apps as possible on whether they are onboard with a change like this, and what everyone's thoughts are on it. Personally I like this structure better, but I haven't been using umbrella apps, so i don't really trust my own opinion on it simply due to lack of experience with them.

**Comment by tsloughter** [comment]
@eproxus  relx does not generate a release that contains all apps under `apps` (our version of an umbrella project), it only includes those that you specify in the `relx` config to be included (+ their deps). So if you have `apps/a` and `apps/b` with relx config `{release, {x, "0.1.0"}, [a]}` and `a` depends on `b` you'll get a release with both `a` and `b`.

If I understood what was being discussed... Very tired but saw this and wanted to reply :)

**Comment by eproxus** [comment]
@tsloughter True. I guess my point is that it doesn't generate one release per app hard coded by default.

I think, I would prefer to have it the relx way. That means, you compose your releases out of selections of apps you choose. The current way in exrm is too limited, in my opinion.

**Comment by bitwalker** [comment]
@eproxus All exrm does today is build a release from the current app, which is exactly what relx does (albeit automated for the most part). In the case of umbrella apps, it treats them as containers of related apps, not necessarily a single logical application, which is how umbrella apps have been treated in mix for quite some time (if not since their inception), and is why exrm iterates over the container and produces releases for each child app. So, if you want an umbrella app to be treated as a single logical app, then we need to change how umbrella apps are handled in exrm, which is not really a problem, however it's something I'd want to get buy in from the community on, since it affects everyone who wants to build a release from an umbrella.

**Comment by tsloughter** [comment]
I'd argue in favor of changing how umbrella apps are treated in exrm. How can we elevate this debate to the community to get buy in?

**Comment by eproxus** [comment]
@bitwalker relx builds a release of the apps you choose in your `relx.config`, not just a single app.

There's no "single logical app" concept in Erlang. There's only apps and releases. Umbrella projects (in Mix and Rebar) are only convenient ways of bundling several apps into one repository (instead of having every app in a sub repo, let's say). What releases you create from those apps is then up to you. That's why I don't think exrm's current default, one release per app, makes any sense, because this is almost never what you want anyway.

**Comment by liveforeverx** [comment]
Seems, like there is a need for exrm.exs for ambrella use cases

**Comment by eproxus** [comment]
Here's my proposal:

### Single App Projects

* `mix release` - exrm builds the release as normal, using the single app as the only included application in the release (together with its dependencies, of course). This is the way exrm works today.
* `mix release /path/to/relx.config` - This would build a custom release according to specifications. For example, including apps that are deps in Mix that the app is not depending on (this could be debugging tools or other utilities that the user wants included in the release).

### Umbrella Projects

* `mix release` - exrm builds a single release under `./rel` (not inside each app) with _all_ apps in the umbrella project. The release is named after the project.
* `mix release [apps]` - This will build a release for only a single app (together with its dependencies). This will allow for multi-node projects having their own apps inside `./apps` together with their dependencies.
* `mix release /path/to/relx.config` - This would build a custom release according to specifications. For example, including apps that are deps in Mix that any of the apps are not depending on (this could be debugging tools or other utilities that the user wants included in the release).


I'm not sure how much magic exrm provides around the `relx.config` file so having that as a direct target might not be appropriate, but since there's no other exrm config file, this was the only thing I could think of. Let me know if there's anything more appropriate to point to.

**Comment by bitwalker** [comment]
@tsloughter The mailing list is the best place really, and I agree.

@eproxus I am aware that the concept of single logical app vs container of apps doesn't make sense in Erlang, I am using those terms to describe how people approach Elixir applications, because mix itself treats single apps vs umbrella apps that way, whether it makes any sense or not is sort of irrelevant to my point. In any case, we're all in agreement here that what you're proposing makes the most sense.

Exrm doesn't do much in terms of magic around the relx.config file, other than generating the defaults. If you put your own in `rel/relx.config`, that file will override any exrm defaults. Where you might run into trouble is if you define your own releases in that config, since that's where exrm does the most "magic".

As far as your proposal goes:

1.) Looks good of course
2.) This already works today by providing `rel/relx.config`. I have not however tried putting custom release definitions in this file, so that may not work (or maybe it does, who knows).
3.) As we've been talking about, I think this makes sense
4.) I assume you mean that this would build a single release containing all the apps in `[apps]` yeah? I think @liveforeverx is not far off the mark in that we'll need to have an exrm configuration file (probably just loaded from `config.exs` though).
5.) More or less the same as 2, except we would have to implement 3 first.

I think a good next step would be to post the proposal, or a link back to this thread, to the mailing list, soliciting feedback from users of exrm. I imagine most people will be on board with this, but it would be good to get as much feedback as possible.

**Comment by tsloughter** [comment]
Ok, let me know if someone starts the convo on the elixir mailing list so I can join :)

**Comment by eproxus** [comment]
I'm not on that list, will join and take a look around. :smile:

**Comment by eproxus** [comment]
Core or talk?

**Comment by tsloughter** [comment]
I joined talk, so hopefully talk :)

Also, I think exrm should just have a config section in config.exs that it uses it create the relx config. This is what rebar3 does, though in rebar3 it simply passes the contents of `{relx, _}` from the config to `relx:main/1` directly.

**Comment by bitwalker** [comment]
elixir-lang-talk for sure :)

@tsloughter Agreed, no reason why we couldn't go that route, just need to convert the terms to Erlang terms when passing it to relx, but I already have code to do that somewhere (conform I think).

**Comment by ulrikaugustsson** [comment]
Is this being worked on? anything I can do to help?

**Comment by bitwalker** [comment]
@ulrikaugustsson It is being worked on, albeit slowly due to other higher priority issues I've been working on. If you'd like to contribute, I can certainly set aside some time to talk over thing on IRC or Slack with you, and I'm usually pretty responsive by email too. If not, that's totally cool, just know it's on my TODO list :)
